db/bbolt.go

```diff
@@ -0,0 +1,262 @@
+		logger: logger,
+	}, nil
+}
+
+//////////////////////////////
+// METHODS
+//////////////////////////////
+
+func (bb *BBoltHandler) CreateInitialSchema() error {
+	bb.logger.Info("Initialising DB...")
+	return bb.db.Update(func(tx *bolt.Tx) error {
+		buckets := []string{blocksBucket, blockHeightsBucket, latestBlockBucket}
+		for _, bucket := range buckets {
+			if err := bb.tryCreateBucket(tx, bucket); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+func (bb *BBoltHandler) InsertBlock(block *types.Block) error {
+	bb.logger.Info("Inserting block to DB", zap.Uint64("block_height", block.BlockHeight))
+
+	// Store mapping number -> block
+	err := bb.db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(blocksBucket))
+		key := bb.itob(block.BlockHeight)
+		blockBytes, err := json.Marshal(block)
+		if err != nil {
+			return err
+		}
+		return b.Put(key, blockBytes)
+	})
+	if err != nil {
+		bb.logger.Error("Error inserting block", zap.Error(err))
+		return err
+	}
+
+	// Store mapping hash -> number
+	err = bb.db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(blockHeightsBucket))
+		return b.Put([]byte(block.BlockHash), bb.itob(block.BlockHeight))
+	})
+	if err != nil {
+		bb.logger.Error("Error inserting block", zap.Error(err))
+		return err
+	}
+
+	// Get current latest block
+	latestBlock, err := bb.GetLatestBlock()
+	if latestBlock == nil {
+		latestBlock = &types.Block{BlockHeight: 0}
+	}
+	if err != nil {
+		bb.logger.Error("Error getting latest block", zap.Error(err))
+		return err
+	}
+
+	// Update latest block if it's the latest
+	err = bb.db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(latestBlockBucket))
+		if err != nil {
+			bb.logger.Error("Error getting latest block", zap.Error(err))
+			return err
+		}
+		if latestBlock.BlockHeight < block.BlockHeight {
+			return b.Put([]byte(latestBlockKey), bb.itob(block.BlockHeight))
+		}
+		return nil
+	})
+	if err != nil {
+		bb.logger.Error("Error updating latest block", zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+func (bb *BBoltHandler) GetBlockByHeight(height uint64) (*types.Block, error) {
+	var block types.Block
+	err := bb.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(blocksBucket))
+		v := b.Get(bb.itob(height))
+		if v == nil {
+			return types.ErrBlockNotFound
+		}
+		return json.Unmarshal(v, &block)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &block, nil
+}
+
+func (bb *BBoltHandler) GetBlockByHash(hash string) (*types.Block, error) {
+	// Fetch block number corresponding to hash
+	var blockHeight uint64
+	err := bb.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(blockHeightsBucket))
+		v := b.Get([]byte(hash))
+		if v == nil {
+			return types.ErrBlockNotFound
+		}
+		blockHeight = bb.btoi(v)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return bb.GetBlockByHeight(blockHeight)
+}
+
+func (bb *BBoltHandler) GetBlockStatusByHeight(height uint64) (bool, error) {
```

Comment:
should we rename it and the method below to QueryIsBlockActiveByHeight? if not, we should at least add some comments to explain what true/false mean

Reply:
should we call it `QueryIsBlockFinalizedByHeight` instead? 'active' seems to be nomenclature used to indicate FPs with voting power so may be a bit confusing

Reply:
ah yeah `QueryIsBlockFinalizedByHeight` is better